cmd/fyne/internal/commands/translate.go

```diff
@@ -0,0 +1,374 @@
+package commands
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"go/ast"
+	"go/parser"
+	"go/token"
+	"io"
+	"io/fs"
+	"os"
+	"path/filepath"
+	"strconv"
+
+	"github.com/urfave/cli/v2"
+)
+
+// Translate returns the cli command to scan for new translation strings.
+func Translate() *cli.Command {
+	return &cli.Command{
+		Name:  "translate",
+		Usage: "Scans for new translation strings.",
+		Flags: []cli.Flag{
+			&cli.BoolFlag{
+				Name:    "verbose",
+				Aliases: []string{"v"},
+				Usage:   "Show files that are being scanned etc.",
+			},
+			&cli.BoolFlag{
+				Name:    "update",
+				Aliases: []string{"u"},
+				Usage:   "Update existing translations (use with care).",
+			},
+			&cli.BoolFlag{
+				Name:    "dry-run",
+				Aliases: []string{"n"},
+				Usage:   "Scan without storing the results.",
+			},
+			&cli.StringFlag{
+				Name:    "sourceDir",
+				Aliases: []string{"d"},
+				Usage:   "Directory to scan recursively for go files.",
+				Value:   ".",
+			},
+			&cli.StringFlag{
+				Name:    "translationsFile",
+				Aliases: []string{"f"},
+				Usage:   "File to read from and write translations to.",
+				Value:   "translations/en.json",
+			},
+		},
+		Action: func(ctx *cli.Context) error {
+			sourceDir := ctx.String("sourceDir")
+			translationsFile := ctx.String("translationsFile")
+			files := ctx.Args().Slice()
+			opts := translateOpts{
+				DryRun:  ctx.Bool("dry-run"),
+				Update:  ctx.Bool("update"),
+				Verbose: ctx.Bool("verbose"),
+			}
+
+			// without any argument find all .go files in the source directory
+			if len(files) == 0 {
+				sources, err := findFilesExt(sourceDir, ".go")
+				if err != nil {
+					return err
+				}
+				files = sources
+			}
+
+			// update the translation file by scanning the given source files
+			return updateTranslationsFile(&opts, translationsFile, files)
+		},
+	}
+}
+
+// Recursively walk the given directory and return all files with the matching extension
+func findFilesExt(dir, ext string) ([]string, error) {
+	files := []string{}
+	err := filepath.Walk(dir, func(path string, fi fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		if filepath.Ext(path) != ext || fi.IsDir() || !fi.Mode().IsRegular() {
+			return nil
+		}
+
+		files = append(files, path)
+
+		return nil
+	})
+	return files, err
+}
+
+type translateOpts struct {
+	DryRun  bool
+	Update  bool
+	Verbose bool
+}
+
+// Create or add to translations file by scanning the given files for translation calls
+func updateTranslationsFile(opts *translateOpts, file string, files []string) error {
+	translations := make(map[string]interface{})
+
+	// try to get current translations first
+	f, err := os.Open(file)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return err
+	}
+	defer f.Close()
```

Comment:
This line will be reached when the file does not exist - seems like it should not be running.

Reply:
It is meant to keep going even if the file doesn't exist. The idea is to make it work without having to create a translation file first. Need to make sure this is mentioned in the documentation, thanks.

Reply:
Another way to look at it: we want to reach this line when there is an error other than the file not existing (something fatal).

Reply:
Sorry I missed this comment. As far as I can tell this code will execute for one of two reasons:
a) The file opened OK (no error)
b) the file could not be opened (does not exist).

For b) we should not be closing the file! I think all that needs to happen is that the defer moves into the `if` below

Reply:
Oh, I think I might have misunderstood. Yes, moving the defer is cleaner..